### DISCUSSION TITLE
[Package] Semantic Types (Typed Entities)

### DISCUSSION THREAD

**Discussion by kyleaton** [great-expectations/great_expectations#4370] [Expectations]
# What Are Semantic Types?	

That's a great question! Semantic Types categorize data by the type of information it represents. For example, consider a column with STRING data types. How do you know what the data means? You could be looking at a list of streets, cities, counties, etc. Without background information, it can be hard to know what you're looking at. That's where Semantic Types come in. The entity’s Semantic Type defines how it should be interpreted.

## Why should we test on Semantic Types?

Testing on these Semantic Types allows us to create more explicit, fit-for-purpose tests, unlocking questions like `Does this column contain US State abbreviations?` instead of asking `Does this column contain strings in the set [AL, AL, AR, AZ...]`

## How do you know when data entities are good candidates for Semantic Types testing? 

Any data that you can tie to a real-world category or reference is ideal for this kind of test, e.g., phone numbers, ZIP codes, countries, coordinates, URLs, email addresses, etc.

Linked below are a number of Expectations on Semantic Types that have already been contributed to Great Expectations!

**Comment by austiezr** [comment]
- [Expect Column Values Are In Language](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_are_in_language.py)
- [Expect Column Values To Be A Non-Bot User Agent](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_a_non_bot_user_agent.py)
- [Expect Column Values To Be ASCII](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_ascii.py)
- [Expect Column Values To Be Secure Passwords](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_secure_passwords.py)
- [Expect Column Values To Be US Zipcode Within Radius Of Given Zipcode](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_us_zipcode_within_radius_of_given_zipcode.py)
- [Expect Column Values To Be Valid Degree-Decimal Coordinates](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_valid_degree_decimal_coordinates.py)
- [Expect Column Values To Be Valid URLs](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_valid_urls.py)
- [Expect Column Values To Be Valid UUID](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_valid_uuid.py)
- [Expect Column Values To Be Valid Wikipedia Articles](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_be_valid_wikipedia_articles.py)
- [Expect Column Values To Contain Valid Email](https://github.com/great-expectations/great_expectations/blob/develop/contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_contain_valid_email.py)

**Reply by joshua-stauffer** [reply]
I'm implementing a metric which calculates an input against the Luhn algorithm, and will contribute `expect_column_values_to_be_valid_luhn_number` and `expect_column_values_to_be_valid_credit_card_number`, which will both use the metric. There are a number of other identifiers which use the algorithm, so I might get around to some of them as well.

**Comment by voidforall** [comment]
Hello, I am new to join the hackathon and really interested in contributing to the Semantic Types!

I have submitted a PR yesterday to [expect column values to be valid IPv4 address](https://github.com/great-expectations/great_expectations/pull/4457). It's a trivial feature and I am also trying to learn the contribution process from it. I guess it could exist many problems and your review can help me understand and fix them. @austiezr @kyleaton

**Comment by austiezr** [comment]
Hey Hackathon People! 

As we come into the final hours of the [Great Expectations Hackathon](https://greatexpectations.devpost.com/), we wanted to briefly highlight what we believe would be some quick wins in the Semantic Types category enabled by [Regex-Based Column Map Expectations](https://docs.greatexpectations.io/docs/guides/expectations/creating_custom_expectations/how_to_create_custom_regex_based_column_map_expectations/) and [Set-Based Column Map Expectations](https://docs.greatexpectations.io/docs/guides/expectations/creating_custom_expectations/how_to_create_custom_set_based_column_map_expectations/).

## Regex-Based Column Map Expectations

- ~~expect_column_values_to_be_[valid_msisdn](https://en.wikipedia.org/wiki/MSISDN)~~
- expect_column_values_to_be_[valid_upc_a](https://en.wikipedia.org/wiki/Universal_Product_Code)
- expect_column_values_to_be_[valid_upc_e](https://en.wikipedia.org/wiki/Universal_Product_Code#UPC-E)
- ~~expect_column_values_to_be_[valid_ean_13](https://en.wikipedia.org/wiki/International_Article_Number)~~
- ~~expect_column_values_to_be_[valid_isbn_10](https://en.wikipedia.org/wiki/International_Standard_Book_Number)~~

## Set-Based Column Map Expectations

- ~~expect_column_values_to_be_[valid_top_level_domain](https://en.wikipedia.org/wiki/Top-level_domain)~~
- ~~expect_column_values_to_be_[valid_http_method](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)~~
- ~~expect_column_values_to_be_[valid_iso_639_1](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)~~
- ~~expect_coolumn_values_to_be_valid_[zipcode_prefixes_in_new_hampshire](https://en.wikipedia.org/wiki/List_of_ZIP_Code_prefixes)~~
- expect_column_values_to_be_valid_[canadian_administrative_divisions](https://en.wikipedia.org/wiki/Provinces_and_territories_of_Canada)

---

Creating expectations of these types can be as simple as identifying a valid regex or value-set and writing a couple of tests; the new `RegexBased` and `SetBasedColumnMapExpectations` will take care of the rest!

**Reply by szecsip** [reply]
@austiezr Thank you for the suggestions. I implemented most barcode validating in one expectation with a third-party package and the users can set the barcode type. (PR #4664)

**Reply by Derekma73** [reply]
@austiezr can you clear up what the intent behind posting the expect_coolumn_values_to_be_valid_zipcode_prefixes_in_new_hampshire as an expectation was? It was my understanding that that level of granularity for an expectation was acceptable with the idea that all custom expectations could be further organized in the contrib folder. When thinking about the end-user, my thought was that it would be easier for them to assign an extremely specific custom expectation to their column rather than spend time configuring the expectation (more clicks now = less clicks for them later).
